### ISSUE TITLE
Mouse sensitivity

### ISSUE THREAD

**Issue by Steven20102** [ValveSoftware/halflife#3334]
Can you guys tell us what you changed with the cs 1.6 mouse input?

I just downloaded the non steam cs 1.6 clean version and use -noforcemparms and -noforcemaccel and the sensitivity feels perfect like in kovaaks (even tho i get random acceleration)

@SamVanheer do you know what is going on? Steam Cs 1.6 version feels now excat like csgo with -useforcedmparms and m_mousespeed 0

**Comment by SamVanheer** [comment]
Steam HL uses SDL2 just like Source and Source 2, whatever old version you tried is pre-SDL so it uses raw mouse input instead.

You can see which changes were made to the SDK mouse input code here: https://github.com/SamVanheer/Half-Life-Tools-Backup/commit/c18aa706f3cb2a4a0332f2dd142d4c99dd017533#diff-e9df30be9a600142970347ccd084b6f3aa4c5dc78516959243ea382d9c5f250d

You can see which options were added to control mouse input, see if any of them help.

**Comment by Steven20102** [comment]
@SamVanheer do you know what they hardcoded for the threshold1 and threshold2?

**Comment by SamVanheer** [comment]
No, both the old and new code uses 0 for both thresholds. That's been the case since Quake 1: https://github.com/id-Software/Quake/blob/bf4ac424ce754894ac8f1dae6a3981954bc9852d/WinQuake/in_win.c#L43

CS 1.6 has the same values.

SDL2 queries Windows the same way to adjust mouse speed: https://github.com/libsdl-org/SDL/blob/c5c94a6be6bfaccec9c41f6326bd4be6b2db8aea/src/video/windows/SDL_windowsmouse.c#L470-L483

It's possible that SDL2 is affecting mouse sensitivity because of this code.

Try disabling mouse acceleration: https://www.makeuseof.com/windows-10-disable-mouse-acceleration/#how-to-disable-mouse-acceleration-feature-in-windows-10-and-11

See if that solves your problem.

**Comment by rikkitikkitavix** [comment]
my english is bad
can you try
-noforcemaccel -noforcemparms -noforcemspd
-noforcemparms -noforcemaccel -noforcemspd
-noforcemspd -noforcemaccel -noforcemparms
different method works differently

**Comment by Steven20102** [comment]
Agree, the different order doesnt change anything. noforcemparms is the same as noforcemspd + noforcemaccel.

_Raw mouse input is supposed to bypass SDL2 and use the values from the OS directly, so that should avoid the issue._

As far as i know raw mouse input is supposed to bypass the windows settings. So the hardcoded values in the engine could still have an effect.

Since CSS and Csgo is customer friendly with their ingame commands m_mousespeed and m_mouseaccel1/2 which are changeable i will test there.

**Comment by rikkitikkitavix** [comment]
in cs1.6 it has different

**Comment by Steven20102** [comment]
No see here the mouse code:

_if (mouseparmsvalid)
	{
		if ( gEngfuncs.CheckParm ("-noforcemspd", NULL ) ) 
			newmouseparms[2] = originalmouseparms[2];

		if ( gEngfuncs.CheckParm ("-noforcemaccel", NULL ) ) 
		{
			newmouseparms[0] = originalmouseparms[0];
			newmouseparms[1] = originalmouseparms[1];
		}

		if ( gEngfuncs.CheckParm ("-noforcemparms", NULL ) ) 
		{
			newmouseparms[0] = originalmouseparms[0];
			newmouseparms[1] = originalmouseparms[1];
			newmouseparms[2] = originalmouseparms[2];_

You can see noforcemparms is the same as noforcemspd and noforcemaccel

**Comment by SamVanheer** [comment]
> Na it seems like they control the windows settings and set windows accel to on. Even when the checkbox in windows is not ticked.
> Or do you mean the have an inbuilt accel system which is the same as windows?

SDL2 queries Windows for the mouse settings before the client has changed those settings, so SDL2 is going off of the OS settings, not the ones set by the client. It then uses those settings to apply adjustments to mouse input so when the game queries SDL2 for the mouse position it will be affected by the OS settings.

> As far as i know raw mouse input is supposed to bypass the windows settings. So the hardcoded values in the engine could still have an effect.

Yeah i got it backwards. Raw input uses SDL2, so turning raw input off should use the OS mouse position.

Source has completely different input code: https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/client/in_mouse.cpp

So i wouldn't count on Source for improvements here.

**Comment by Steven20102** [comment]
damn its so annoying dealing with different mouse feeling even if they have (near) the same engine. Iam not sure how to make the mouse the same in cs 1.6 non steam and steam version. Also I cant find a way (for now) to make source games like 1.6 non steam.

But i also think the problem is that windows is still influencing the mouse input.

**Comment by rikkitikkitavix** [comment]
just try these variations on practice

**Comment by Steven20102** [comment]
I found this: http://web.archive.org/web/20090923154337/http://www.gotfrag.com/cs/forums/thread/411262

_#39 Even though you may have received that from technical support at [steampowered.com](http://web.archive.org/web/20090923154337/http://steampowered.com/) I believe it is incorrect, because the code from the half-life sdk disagrees with it. The only part they have correct is -noforcemaccel, but they fail to mention that it only applies if you have acceleration turned on. In other words, this setting only makes Half-Life use the acceleration thresholds you have set in Windows, but if you have acceleration turned off, this is ignored.

#40's code snippet is the same one that I examined, from the half-life sdk, that my original thread is based upon. It's the code that initializes and activates the mouse.

#41 You are seeing the technical proof of my original post in #40's code snippet. Frankly I think you may be wrong here, because from a development standpoint, there's little reason to create a command that slows the y-axis of a mouse. And as the code shows, -noforcemspd just doesn't do what you say it does.

One possible reason I can think of for the difference in feeling that you are experiencing (and this is a long shot) is that the mouse initialization code might be called more frequently than I expect. If it is called frequently, then having more -noforce commands specified makes that sub-routine longer (though marginally).

In other words, if you have just -noforcemparms specified, you only execute through the bottom part of #40's code snippet, whereas if you have all three, you execute through the whole function. It's redundant though, essentially the function you're executing becomes twice as long. Though we're talking milliseconds here, so again, it's probably negligible._

So maybe you are right. I dont know gldscr and source engine is really weird. Actually ppl in 2009 discussed this problems.

**Comment by Steven20102** [comment]
```
void WIN_UpdateMouseSystemScale()
{
    int mouse_speed;
    int params[3] = { 0, 0, 0 };

    if (SystemParametersInfo(SPI_GETMOUSESPEED, 0, &mouse_speed, 0) &&
        SystemParametersInfo(SPI_GETMOUSE, 0, params, 0)) {
        if (params[2]) {
            WIN_SetEnhancedMouseScale(mouse_speed);
        } else {
            WIN_SetLinearMouseScale(mouse_speed);
        }
    }
}
```

So we know cs (all cs even source) use params[3] = { 0, 0, 1 }

We know the first two 0 reffering to mouse threshold and the last to the mouse speed.
The Code turns mouse speed always on by default. Only noforcemspd which is reffering to "params[2]" change that. With noforcemspd you tell the engine that windows set the last parameter. So if you use 0 in windows its now: params[3] = { 0, 0, 0 }

Short: Set in all goldscr or source games noforcemspd in the launch options. Yes even in source it works without the need of useforcedmparms

**Comment by rikkitikkitavix** [comment]
can you check if old nonsteam cs 1.6 using cl_mousegrab 1 or similar thing

**Comment by Steven20102** [comment]
After 1 month testing i have to say you right. They feel different and also the order of the launch options affects it.

I guess right is just -noforcemparms or -noforcemaccel -noforcemspd

both of this will give params[3] = { 0, 0, 0 } and in the right order.

**Comment by rikkitikkitavix** [comment]
what i should use m_customaccel_exponent 1 or 0 and why?

**Comment by rikkitikkitavix** [comment]
in cs many values that should not affect mouse sensitivity affect this
as well as integer values and dotted values work differently. I don't know how.
for example sensitvity 3 and 3.0 or 3.000000 are felt differently
zoom_sensitivity_ratio 1 and 1.0
try lookspring 0.000000 and 0
lookstrafe 0.000000 and 0

**Comment by rikkitikkitavix** [comment]
after changing the value, it is better to restart the game
